crates/bevy_math/src/primitives/dim2.rs

```diff
@@ -0,0 +1,134 @@
+use super::Primitive2d;
+use crate::Vec2;
+
+/// A normalized vector pointing in a direction in 2D space
+#[derive(Clone, Copy, Debug)]
+pub struct Direction2d(Vec2);
+
+impl From<Vec2> for Direction2d {
+    fn from(value: Vec2) -> Self {
+        Self(value.normalize())
+    }
+}
+
+impl Direction2d {
+    /// Create a direction from a [Vec2] that is already normalized
+    pub fn from_normalized(value: Vec2) -> Self {
+        Self(value)
+    }
+}
+
+impl std::ops::Deref for Direction2d {
+    type Target = Vec2;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// An infinite half-line pointing in a direction in 2D space
+#[derive(Clone, Copy, Debug)]
+pub struct Ray2d(pub Direction2d);
+
+/// A circle primitive
+#[derive(Clone, Copy, Debug)]
+pub struct Circle {
+    /// The radius of the circle
+    pub radius: f32,
+}
+impl Primitive2d for Circle {}
+
+/// An unbounded plane in 2D space
+#[derive(Clone, Copy, Debug)]
+pub struct Plane2d {
+    /// The direction in which the plane points
+    pub normal: Direction2d,
+}
+impl Primitive2d for Plane2d {}
+
+/// An infinite line along a direction in 2D space.
+/// For a finite line: [`LineSegment2d`]
+#[derive(Clone, Copy, Debug)]
+pub struct Line2d {
+    /// The direction of the line
+    pub direction: Direction2d,
+}
+impl Primitive2d for Line2d {}
+
+/// A section of a line along a direction in 2D space.
+#[derive(Clone, Debug)]
+pub struct LineSegment2d {
+    /// The direction of the line
+    pub direction: Direction2d,
+    /// The point where the line starts
+    pub start: f32,
+    /// The point where the line ends
+    pub end: f32,
+}
+impl Primitive2d for LineSegment2d {}
+
+/// A line alone a path of N vertices in 2D space.
+/// For a version without generics: [`BoxedPolyline2d`]
+#[derive(Clone, Debug)]
+pub struct Polyline2d<const N: usize> {
+    /// The vertices of the polyline
+    pub vertices: [Vec2; N],
+}
+impl<const N: usize> Primitive2d for Polyline2d<N> {}
+
+/// A line alone a path of vertices in 2D space.
+/// For a version without alloc: [`Polyline2d`]
+#[derive(Clone, Debug)]
+pub struct BoxedPolyline2d {
+    /// The vertices of the polyline
+    pub vertices: Box<[Vec2]>,
```

Comment:
What are the pros and cons of using this versus a `Vec<Vec2>`?

Reply:
Box<[]> has a smaller memory footprint since it's just a length vs length and capacity yea. Might even be worth to have a Vec, Box AND const generics option, tho at that point they'd probably need to be aliases to a generic version :thinking: